### PR TITLE
Support Twitter futures

### LINKED ISSUE #718: Support Twitter Futures
Hi!
Are there any plans to add support for [Twitter Futures](https://twitter.github.io/util/guide/util-cookbook/futures.html)?

### PR DESCRIPTION
**Pull request by mijicd**
Closes #718.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -36,7 +36,7 @@ jobs:
     - &scaladoc
       stage: microsite
       name: "Generate Scaladoc"
-      script: sbt ++$TRAVIS_SCALA_VERSION coreJVM/doc coreJS/doc interopCatsJVM/doc interopCatsJS/doc interopFutureJVM/doc interopFutureJS/doc interopScalaz7xJVM/doc interopScalaz7xJS/doc interopMonixJVM/doc interopMonixJS/doc interopJavaJVM/doc interopReactiveStreamsJVM/doc
+      script: sbt ++$TRAVIS_SCALA_VERSION coreJVM/doc coreJS/doc interopCatsJVM/doc interopCatsJS/doc interopFutureJVM/doc interopScalaz7xJVM/doc interopScalaz7xJS/doc interopMonixJVM/doc interopMonixJS/doc interopJavaJVM/doc interopReactiveStreamsJVM/doc interopTwitterJVM/doc
```
Comment (LGLO) [inline .travis.yml]: is `interopFutureJS/doc` gone by accident or for purpose? WHy?

File: build.sbt [hunk 2]
```diff
@@ -42,13 +42,14 @@ lazy val root = project
...
-//  interopMonixJS,
+    interopMonixJVM,
+    interopMonixJS,
     interopScalaz7xJVM,
     interopScalaz7xJS,
     interopJavaJVM,
     interopReactiveStreamsJVM,
-//  benchmarks,
+    interopTwitterJVM,
+    benchmarks,
```
Comment (LGLO) [inline build.sbt]: What was the reason to have `benchmarks` commented out?

File: .travis.yml (hunk 1 above)
Comment (mijicd) [reply .travis.yml]: It was on purpose, as I think there isn't (wasn't?) such interop atm. I can reintroduce it though.

File: build.sbt (hunk 2 above)
Comment (mijicd) [reply build.sbt]: No idea. It was commented out for a long time now. As far as I can see it's commented out together with Monix during "death of TF" PR :).

**Comment by mijicd** [comment]
@LGLO Thanks for the review. I will add the tests.

File: docs/interop/twitter.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+---
+id: interop_twitter
```
Comment (LGLO) [inline docs/interop/twitter.md]: As we moved to docusaurus you could add reference to this in `website/sidebar.json`

**Comment by mijicd** [comment]
@LGLO added docs to sidebar.

File: interop-twitter/jvm/src/main/scala/scalaz/zio/interop/twitter.scala [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package scalaz.zio.interop
+
+import com.twitter.util.{ Future, Return, Throw }
+import scalaz.zio.Task
+
+package object twitter {
+  implicit class TaskObjOps(private val obj: Task.type) extends AnyVal {
+    final def fromTwitterFuture[A](future: => Future[A]): Task[A] =
+      Task.effectAsync { cb =>
```
Comment (regiskuckaertz) [inline interop-twitter/jvm/src/main/scala/scalaz/zio/interop/twitter.scala]: You will also need to handle interruption from the outside with `future.raise`

File: build.sbt (hunk 2 above)
Comment (wi101) [reply build.sbt]: I remember it :D , I wondered why it was already commented before

File: interop-twitter/jvm/src/main/scala/scalaz/zio/interop/twitter.scala (hunk 4 above)
Comment (mijicd) [reply interop-twitter/jvm/src/main/scala/scalaz/zio/interop/twitter.scala]: Any suggestions on the approach? From what I see, interrupt handler is available only for `Promise`, and raising is propagated to producers.

Comment (yaroot) [reply interop-twitter/jvm/src/main/scala/scalaz/zio/interop/twitter.scala]: I've made some change [here](https://github.com/yaroot/scalaz-zio/commits/twitter-future)

Interruption handler in promise is where you need to accept downstream cancellation, here you need to signal cancellation by using `future.raise`.

Besides that `fromTwitterFuture` should really accept `Task[Future[A]]` (similar to the scala version `fromFuture`).

Just my 2 cents

Comment (regiskuckaertz) [reply interop-twitter/jvm/src/main/scala/scalaz/zio/interop/twitter.scala]: As @yaroot did 😄

Comment (mijicd) [reply interop-twitter/jvm/src/main/scala/scalaz/zio/interop/twitter.scala]: @yaroot `fromFuture` doesn't accept `Task`, as far as I can see, it accepts `ExecutionContext => scala.concurrent.Future[A]`. I'll take a look at the solution you suggested, or, I can retract the PR and let you submit the solution, if it's the preferred approach :).

**Comment by mijicd** [comment]
@LGLO @regiskuckaertz in the end I adjusted the interface as suggested above by @yaroot. Take one more look, and hopefully the build will be green (this flaky reactive test is kinda annoying).

**Pull request merged**